lib/presentation/redux/states/workgroup_nodes_surfing_state.dart

```diff
@@ -0,0 +1,109 @@
+// infringing Linagora intellectual property rights over its trademarks and commercial
+// brands. Other Additional Terms apply, see
+// <http://www.linshare.org/licenses/LinShare-License_AfferoGPL-v3.pdf>
+// for more details.
+// This program is distributed in the hope that it will be useful, but WITHOUT ANY
+// WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+// FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License for
+// more details.
+// You should have received a copy of the GNU Affero General Public License and its
+// applicable Additional Terms for LinShare along with this program. If not, see
+// <http://www.gnu.org/licenses/> for the GNU Affero General Public License version
+//  3 and <http://www.linshare.org/licenses/LinShare-License_AfferoGPL-v3.pdf> for
+//  the Additional Terms applicable to LinShare software.
+
+import 'package:dartz/dartz.dart';
+import 'package:domain/domain.dart';
+import 'package:domain/src/state/failure.dart';
+import 'package:domain/src/state/success.dart';
+import 'package:linshare_flutter_app/presentation/redux/states/linshare_state.dart';
+import 'package:linshare_flutter_app/presentation/util/data_structure/stack.dart';
+
+class WorkGroupNodesSurfingNavigatorState extends LinShareState {
+  WorkGroupNodesSurfingNavigatorState(
+    Either<Failure, Success> viewState,
+    this.nodesHierarchy,
+  ) : super(viewState);
+
+  factory WorkGroupNodesSurfingNavigatorState.initial() {
+    return WorkGroupNodesSurfingNavigatorState(Right(IdleState()), Stack());
+  }
+
+  final Stack<WorkGroupNodesSurfingState> nodesHierarchy;
+
+  WorkGroupNodesSurfingState get currentNodeState => nodesHierarchy.peek();
+
+  @override
+  LinShareState clearViewState() {
+    return WorkGroupNodesSurfingNavigatorState(Right(IdleState()), Stack());
+  }
+
+  @override
+  LinShareState sendViewState({Either<Failure, Success> viewState}) {
+    return WorkGroupNodesSurfingNavigatorState(viewState, nodesHierarchy);
+  }
+
+  LinShareState openNode(WorkGroupNode node) {
+    return WorkGroupNodesSurfingNavigatorState(
+      viewState,
+      nodesHierarchy..push(WorkGroupNodesSurfingState(node, [])),
+    );
+  }
+
+  LinShareState setChildrenCurrentNode(WorkGroupNode node, List<WorkGroupNode> children) {
+    final currentNode = nodesHierarchy.peek().node;
+    if (currentNode.workGroupNodeId == node.workGroupNodeId) {
+      return WorkGroupNodesSurfingNavigatorState(
+        viewState,
+        nodesHierarchy..replaceLast(WorkGroupNodesSurfingState(node, children)),
+      );
+    } else {
+      return this;
+    }
+  }
+
+  LinShareState closeNode() {
+    return WorkGroupNodesSurfingNavigatorState(
+      viewState,
+      nodesHierarchy..pop(),
+    );
+  }
+
+  @override
+  LinShareState startLoadingState() {
+    return WorkGroupNodesSurfingNavigatorState(
+      viewState,
+      nodesHierarchy
+        ..replaceLast(WorkGroupNodesSurfingState(
```

Comment:
can you guide me `..`

Reply:
We can simple think it is type of Builder. After perform, it returned `nodesHierarchy` instance.